lib/accessmonitoring/monitor.go

```diff
@@ -0,0 +1,287 @@
+/*
+ * Teleport
+ * Copyright (C) 2025 Gravitational, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package accessmonitoring
+
+import (
+	"context"
+	"errors"
+	"log/slog"
+	"slices"
+	"time"
+
+	"github.com/gravitational/trace"
+	"golang.org/x/sync/semaphore"
+
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/api/utils/retryutils"
+	"github.com/gravitational/teleport/lib/backend"
+)
+
+const (
+	// componentName specifies the access monitor component name used for debugging.
+	componentName = "access_monitor"
+
+	// maxAccessRequestConcurrency specifies the maximum number of concurretly
+	// handled access request events.
+	maxAccessRequestConcurrency = 20
+	// maxAccessRequestDeadline specifies the maximum time allowed to handle an
+	// access request event.
+	maxAccessRequestDeadline = 10 * time.Second
+
+	// initWatcherTimeout specifies the maximum time to wait for the watcher to
+	// initialize.
+	initWatcherTimeout = 15 * time.Second
+)
+
+// EventHandler describes a function that can handle an access event.
+type EventHandler func(ctx context.Context, event types.Event) error
+
+// Config specifies the access monitor configuration.
+type Config struct {
+	// Logger is the logger for the monitor.
+	Logger *slog.Logger
+
+	// Backend should be a backend.Backend which can be used for obtaining the
+	// lock required to run the service.
+	Backend backend.Backend
+
+	// Events is the event monitor. This interface allows us to monitor access
+	// events.
+	Events types.Events
+
+	// FailFast mode indicates that the access monitor should fail fast instead
+	// of restarting indefinitely. Fail fast mode should only be used within a
+	// plugin environment. See https://github.com/gravitational/teleport/pull/30039
+	// for more details.
+	// TODO(bernardjkim): Investigate if fail fast mode is still required.
```

Comment:
Is this TODO still relevant?

Reply:
Yeah, Hugo raised this concern in https://github.com/gravitational/teleport/pull/53769#discussion_r2038127739.

The plan is to eventually replace the access request plugin `Apps` with this `AccessMonitor` implementation. I'll investigate further at that time if we still need to support a fail fast mode.